raft/src/main/java/org/apache/kafka/raft/LeaderState.java

```diff
@@ -607,20 +565,27 @@ private void updateVoterAndObserverStates(VoterSet lastVoterSet) {
 
         // Move any of the remaining old voters to observerStates
         for (ReplicaState replicaStateEntry : oldVoterStates.values()) {
+            replicaStateEntry.listeners = Optional.empty();
             observerStates.putIfAbsent(replicaStateEntry.replicaKey, replicaStateEntry);
         }
     }
 
-    private static class ReplicaState implements Comparable<ReplicaState> {
+    static class ReplicaState implements Comparable<ReplicaState> {
         ReplicaKey replicaKey;
+        Optional<Endpoints> listeners;
         Optional<LogOffsetMetadata> endOffset;
         long lastFetchTimestamp;
         long lastFetchLeaderLogEndOffset;
         long lastCaughtUpTimestamp;
         boolean hasAcknowledgedLeader;
 
         public ReplicaState(ReplicaKey replicaKey, boolean hasAcknowledgedLeader) {
+            this(replicaKey, hasAcknowledgedLeader, Optional.empty());
+        }
+
+        public ReplicaState(ReplicaKey replicaKey, boolean hasAcknowledgedLeader, Optional<Endpoints> listeners) {
```

Comment:
The `Endpoints` type can be empty. Did you consider using `Endpoints.empty()` instead of `Optional.empty()`?

Reply:
I see, I missed that we will always expect `Endpoints` to be instantiated (an be empty, but not null)